### PR TITLE
[SOFT-435] Setup CI/CD

### PR DESCRIPTION
**Pull request by raiyansayeed**
Opened PR for final debugging of CI

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+      
+      #- name: Setup Mongo
+      # - run: wget -qO - https://www.mongodb.org/static/pgp/server-4.4.asc | sudo apt-key add -
```
Comment (AveryChiu64) [inline .github/workflows/main.yml]: Why are all these commands commented out?

Comment (AveryChiu64) [reply .github/workflows/main.yml]: oh wait actually we were talking about this right? The order that these commands run seems to be mixed up or something? Can we see whats up with that? i'm not sure why these commands would run in an arbitrary order

File: Procfile [hunk 2]
```diff
@@ -0,0 +1,4 @@
+django: cd api && pipenv run start
+mock_data: cd api && pipenv run mock_data
+worker: cd api && export DJANGO_SETTINGS_MODULE=django_app.settings && pipenv run worker
+beat: cd api && export DJANGO_SETTINGS_MODULE=django_app.settings && pipenv run beat
```
Comment (AveryChiu64) [inline Procfile]: Add a newline here

File: README.md [hunk 3]
```diff
@@ -1,20 +1,26 @@
...
-The Django app requires Python >= 3.6 and MongoDB >= 3.6.3
+The Django app requires Python >= 3.6, Pipenv, MongoDB >= 3.6.3 and Redis >= 6.0
 Python3 should be installed in the Vagrant box already, but if you don't have MongoDB...
 
-## MongoDB setup
-To setup mongoDB for data collection run (only necessary the first time you start working with the database)
+## Dependencies setup (Mongo, Pipenv, Redis)
+To setup all dependencies for data collection run (only necessary the first time you start working with the database)
 ```
 bash setup.sh
```
Comment (AveryChiu64) [inline README.md:21]: Let's make sure this setup.sh script is good before we remove everything else below from the readme

File: api/can_server/consumers.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+class ConvertConsumer(WebsocketConsumer):
+    def websocket_connect(self, event):
+        print('connected', event)
+        async_to_sync(self.channel_layer.group_add)(
+            "converted", self.channel_name)
+        self.accept()
+
+    def websocket_receive(self, event):
+        # print(type(event))
```
Comment (AveryChiu64) [inline api/can_server/consumers.py]: Why is there this commented out code?

File: api/README.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
+# Django setup
```
Comment (AveryChiu64) [inline api/README.md:1]: Isn't Django set up already after following the instructions from the readme on the main page? Do we need this readme here?

File: api/can_server/tasks.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+from datetime import datetime
+import json
+import os
+from asgiref.sync import async_to_sync
+from channels.layers import get_channel_layer
+from celery import shared_task
+# from ..django_app.celery import app
+import asyncio
+
+# from celery.task.schedules import crontab
```
Comment (AveryChiu64) [inline api/can_server/tasks.py]: Can we remove these lines that are not needed to keep our code clean?

File: api/can_server/templates/frontend.html [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    convert_socket.onmessage = function(e) {
+      if (should_convert)
+        document.getElementById("can_msg").textContent=e.data;
+
+      var obj = JSON.parse(e.data);
+
+      //console.log(e);
+      //console.log(obj);
```
Comment (AveryChiu64) [inline api/can_server/templates/frontend.html]: Remove this commented out code

File: api/django_app/routing.py [hunk 8]
```diff
@@ -0,0 +1,23 @@
+# from channels.routing import route
+from can_server.consumers import ConvertConsumer
+from channels.routing import ProtocolTypeRouter, URLRouter
+from django.conf.urls import url
+from channels.security.websocket import AllowedHostsOriginValidator
+
+
+# channel_routing = {
```
Comment (AveryChiu64) [inline api/django_app/routing.py]: remove

File: api/django_app/celery.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+@app.task(bind=True)
+def debug_task(self):
+    print('Request: {0!r}'.format(self.request))
+
+
+@app.on_after_configure.connect
+def task_decode_send(sender, **kwargs):
+    sender.add_periodic_task(3, decode_and_send, name='decode_send')
+    # decode_and_send()
+    # logger.info("Sent CAN data")
```
Comment (AveryChiu64) [inline api/django_app/celery.py]: remove unneded code

File: api/system_can.dbc [hunk 10]
```diff
@@ -0,0 +1,317 @@
+VERSION ""
+
+
```
Comment (AveryChiu64) [inline api/system_can.dbc:3]: I would probably put this in the git ignore at some point, we could keep this for now but the user should import the dbc file.

File: client/README.md [hunk 11]
```diff
@@ -0,0 +1,41 @@
+## Mongo Display Setup
```
Comment (AveryChiu64) [inline client/README.md:1]: Do we need this readme as well?

File: setup.sh [hunk 12]
```diff
@@ -12,12 +14,75 @@ echo "mongodb-org-shell hold" | sudo dpkg --set-selections
...
...redis-server /etc/redis/redis.conf\nExecStop=/bin/kill -s TERM $MAINPID\nExecStartPost=/bin/sh -c 'echo $MAINPID > /var/run/redis/redis.pid'\nPIDFile=/run/redis/redis-server.pid\nTimeoutStopSec=0\nRestart=always\nUser=redis\nGroup=redis\nRuntimeDirectory=redis\nRuntimeDirectoryMode=2755\n\nUMask=007\nPrivateTmp=yes\nLimitNOFILE=65535\nPrivateDevices=yes\nProtectHome=yes\nReadOnlyDirectories=/\nReadWriteDirectories=-/var/lib/redis\nReadWriteDirectories=-/var/log/redis\nReadWriteDirectories=-/var/run/redis\n\nNoNewPrivileges=true\nCapabilityBoundingSet=CAP_SETGID CAP_SETUID CAP_SYS_RESOURCE\nRestrictAddressFamilies=AF_INET AF_INET6 AF_UNIX\nMemoryDenyWriteExecute=true\nProtectKernelModules=true\nProtectKernelTunables=true\nProtectControlGroups=true\nRestrictRealtime=true\nRestrictNamespaces=true\n\n# redis-server can write to its own config file when in cluster mode so we\n# permit writing there by default. If you are not using this feature, it is\n# recommended that you replace the following lines with "ProtectSystem=full".\nProtectSystem=true\nReadWriteDirectories=-/etc/redis\n\n[Install]\nWantedBy=multi-user.target\nAlias=redis.service" | sudo tee /etc/systemd/system/redis.service
```
Comment (AveryChiu64) [inline setup.sh]: Why are we printing this?

File: api/README.md (hunk 5 above)
Comment (raiyansayeed) [reply api/README.md:1]: The setup file installs pipenv and node, but developers still have to run ```pipenv install``` and ```npm install``` in the respective api and client libraries. Although I can try putting this in the build script if you don't think it adds more complexity?

File: setup.sh (hunk 12 above)
Comment (raiyansayeed) [reply setup.sh]: It's actually a command to replace the default redis setup. ```sed``` wasn't powerful enough to replace the texts so I had to do a manual approach of adding three lines in the config file

File: client/README.md (hunk 11 above)
Comment (raiyansayeed) [reply client/README.md:1]: It details the specific steps for running the frontend / solving any possible bugs you might encounter. We could add this to the root README, though the latter might get really long.

**Comment by AveryChiu64** [comment]
Hmm ideally I would like the dbc file not to be included with the repository. However, since this will cause the test to fail we leave it in for now.
image

**Pull request closed without merging**